lib/pbench/test/unit/server/query_apis/test_datasets_publish.py

```diff
@@ -83,6 +83,16 @@ def _setup(self, client):
             payload={"controller": "node", "name": "drb", "access": "public"},
         )
 
+    @pytest.mark.parametrize(
+        "malformed_token", ("malformed", "bear token" "Bearer malformed"),
+    )
+    def test_malformed_authorization_header(
+        self, client, server_config, malformed_token, attach_dataset
+    ):
+        super().test_malformed_authorization_header(
+            client, server_config, malformed_token
+        )
```

Comment:
This feels wrong.

How is invoking `TestDatasetsPublish.test_malformed_authorization_header()` different from invoking `Commons.test_malformed_authorization_header()`?  They look like they have the same parameterizations (although, I have to say, I don't know what happens when one parameterized test invokes another...does the "inner" test's parameterization get ignored?); the one invokes the other, so it seems like they have the same "bodies" (do they test the same stuff?); the only difference I note is that the "outer" one is including an additional fixture (but, I can't see _why_, since it doesn't explicitly reference it).

What happens if you simply omit the definition of `TestDatasetsPublish.test_malformed_authorization_header()` and let the class _inherit_ `Commons.test_malformed_authorization_header()`?  Does it not run, or does it encounter some error due to the missing fixture?

Given how much magic PyTest uses, I have a knee-jerk reluctance to having one test function invoke another.  If simple inheritance won't work, then perhaps it would be better to make `Commons.test_malformed_authorization_header()` into a "helper" function (instead of a "test") which `TestDatasetsPublish.test_malformed_authorization_header()` could then simply call (instead of having to make a `super()` reference).

Reply:
Yes when one parameterized test invokes another parameterized test the inner test parametrization gets ignored (because its not called with python decorator). The inner function just acts as any other function as if it didn't have the decorators in the first place. It is certainly possible to make it a helper function in the commons but then it has to be invoked in all the tests where `Commons` is getting inherited.  



> What happens if you simply omit the definition of `TestDatasetsPublish.test_malformed_authorization_header()` and let the class _inherit_ `Commons.test_malformed_authorization_header()`? Does it not run, or does it encounter some error due to the missing fixture?

Yes it encounters error due to the missing fixture because its not being called with the pytest decorator that supplies that fixture value. So we have to explicitly provide the value of the fixture from the function that overriding it.

Reply:
The publish API authorizes against a Dataset rather than against a JSON "user" parameter, requiring the extra fixture.

Now, the Commons version could be progressively edited to pull in every fixture required by an of the subclasses, but that's ugly, too.

While I agree that using the "base class" (undecorated) as a helper is a bit weird, but it seems perfectly legitimate, and it's a pattern to allow the subclasses to specialize their behavior while still leveraging the common code. Within the framework of pytest I don't really have any better ideas.

Reply:
Dave, Nikhil, and I met to discuss this.  We concluded that the least of evils will be to use just the base class method, and to endow it with all the of the fixtures that any of the subclasses will require...at least until we can come up with something better.